iac-extensions/docker/src/main/java/org/sonar/iac/docker/checks/ArgDefinedOutsideOfScopeCheck.java

```diff
@@ -0,0 +1,81 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.iac.docker.checks;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.sonar.check.Rule;
+import org.sonar.iac.common.api.checks.CheckContext;
+import org.sonar.iac.common.api.checks.IacCheck;
+import org.sonar.iac.common.api.checks.InitContext;
+import org.sonar.iac.docker.tree.api.ArgInstruction;
+import org.sonar.iac.docker.tree.api.Body;
+import org.sonar.iac.docker.tree.api.DockerImage;
+import org.sonar.iac.docker.tree.api.Expression;
+import org.sonar.iac.docker.tree.api.Literal;
+import org.sonar.iac.docker.tree.api.RunInstruction;
+import org.sonar.iac.docker.tree.api.Variable;
+
+@Rule(key = "S6579")
+public class ArgDefinedOutsideOfScopeCheck implements IacCheck {
+  private static final String MESSAGE = "Include the ARG instruction in the build stage where it is used.";
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(Body.class, ArgDefinedOutsideOfScopeCheck::checkReferencedVariablesAreInScope);
+  }
+
+  private static void checkReferencedVariablesAreInScope(CheckContext ctx, Body body) {
+    List<String> globalArgsNames = collectArgNames(body.globalArgs().stream());
+
+    body.dockerImages().forEach(image -> checkArgUsagesIn(image, globalArgsNames, ctx));
+  }
+
+  private static List<String> collectArgNames(Stream<ArgInstruction> args) {
+    return args.flatMap(it -> it.keyValuePairs().stream())
+      .map(pair -> {
+        Expression e = pair.key().expressions().get(0);
+        return ((Literal) e).value();
+      })
+      .collect(Collectors.toList());
+  }
+
+  private static void checkArgUsagesIn(DockerImage image, List<String> globalArgNames, CheckContext ctx) {
+    List<String> argNamesInStage = collectArgNames(image.instructions().stream().filter(ArgInstruction.class::isInstance).map(ArgInstruction.class::cast));
+    List<String> notRedeclaredArgsNames = new ArrayList<>(globalArgNames);
+    notRedeclaredArgsNames.removeAll(argNamesInStage);
+
+    List<Variable> usedVariables = image.instructions().stream()
+      .filter(RunInstruction.class::isInstance)
+      .flatMap(run -> ((RunInstruction) run).arguments().stream())
```

Comment:
Should we only check `RunInstruction`? What about references in `CmdInstruction` or somewhere else?

Reply:
I wasn't able to find a comprehensive list of where ARGs can be used. The [docs](https://docs.docker.com/engine/reference/builder/#arg) say

> You can use an ARG or an ENV instruction to specify variables that are available to the RUN instruction.

However, they show example also with `USER`. I think at least it should be covered for all kinds of `CommandInstruction`

Reply:
But I guess the logic for everything other than `FROM` is the same so this can be indeed generalized